theseus/geometry/se3.py

```diff
@@ -0,0 +1,167 @@
+            self._SE3_matrix_check(data)
+        super().__init__(data=data, name=name, dtype=dtype)
+        if x_y_z_quaternion is not None:
+            if x_y_z_quaternion is not None:
+                if x_y_z_quaternion.ndim == 1:
+                    x_y_z_quaternion = x_y_z_quaternion.unsqueeze(0)
+                self.update_from_x_y_z_quaternion(x_y_z_quaternion=x_y_z_quaternion)
+
+    @staticmethod
+    def _init_data() -> torch.Tensor:  # type: ignore
+        return torch.eye(3, 4).view(1, 3, 4)
+
+    def dof(self) -> int:
+        return 6
+
+    def __repr__(self) -> str:
+        return f"SE3(data={self.data}, name={self.name})"
+
+    def __str__(self) -> str:
+        with torch.no_grad():
+            return f"SE3(matrix={self.data}), name={self.name})"
+
+    def _adjoint_impl(self) -> torch.Tensor:
+        ret = torch.zeros(self.shape[0], 6, 6).to(dtype=self.dtype, device=self.device)
+        ret[:, :3, :3] = self[:, :3, :3]
+        ret[:, 3:, 3:] = self[:, :3, :3]
+        ret[:, :3, 3:] = SO3.hat(self[:, :3, 3]) @ self[:, :3, :3]
+
+        return ret
+
+    @staticmethod
+    def _SE3_matrix_check(matrix: torch.Tensor):
+        if matrix.ndim != 3 or matrix.shape[1:] != (3, 4):
+            raise ValueError("SE(3) can only be 3x4 matrices.")
+        SO3._SO3_matrix_check(matrix.data[:, :3, :3])
+
+    @staticmethod
+    def _hat_matrix_check(matrix: torch.Tensor):
+        if matrix.ndim != 3 or matrix.shape[1:] != (4, 4):
+            raise ValueError("Hat matrices of SE(3) can only be 4x4 matrices")
+
+        if matrix[:, 3].abs().max().item() > theseus.constants.EPS:
+            raise ValueError("The last row of hat matrices of SE(3) can only be zero.")
+
+        if (
+            matrix[:, :3, :3].transpose(1, 2) + matrix[:, :3, :3]
+        ).abs().max().item() > theseus.constants.EPS:
+            raise ValueError(
+                "The 3x3 top-left corner of hat matrices of SE(3) can only be skew-symmetric."
+            )
+
+    @staticmethod
+    def exp_map(tangent_vector: torch.Tensor) -> LieGroup:
+        raise NotImplementedError
+
+    def _log_map_impl(self) -> torch.Tensor:
+        raise NotImplementedError
+
+    def _compose_impl(self, so3_2: LieGroup) -> "SE3":
+        raise NotImplementedError
+
+    def _inverse_impl(self, get_jacobian: bool = False) -> "SE3":
+        ret = torch.zeros(self.shape[0], 3, 4).to(dtype=self.dtype, device=self.device)
+        ret[:, :, :3] = self.data[:, :3, :3].transpose(1, 2)
+        ret[:, :, 3] = -(ret[:, :3, :3] @ self.data[:, :3, 3].unsqueeze(2)).view(-1, 3)
+        return SE3(data=ret)
+
+    def to_matrix(self) -> torch.Tensor:
+        ret = torch.zeros(self.shape[0], 4, 4).to(dtype=self.dtype, device=self.device)
+        ret[:, :3] = self.data
+        ret[:, 3, 3] = 1
+        return ret
+
+    def update_from_x_y_z_quaternion(self, x_y_z_quaternion: torch.Tensor):
+        if x_y_z_quaternion.ndim != 2 and x_y_z_quaternion.shape[1] != 7:
```

Comment:
We want to handle the case where the batch dimension is not passed and therefore batch size is assumed to be 1 (possibly also check in other places if we account for this).

Reply:
OK. I will add this feature.

Reply:
TBH, I'm not sure if we support this in 2D functions. Should we open an issue to remember to check for this?

Reply:
There is `SO2.update_from_angle` but `SE2` seems to have no similar functions.

Reply:
> TBH, I'm not sure if we support this in 2D functions. Should we open an issue to remember to check for this?

Okay with either of these but we should be consistent:
a. We clarify in docs that batch dimension is always present and users need to add a stub dim for a batch size of one.
b. We let users define things without batch and append the extra dim internally everywhere.

Reply:
Actually, for some reason I was thinking of something else. In Lie group classes constructors we do automatically append batch dimensions if not present, and @fantaosha also added this here in [line 35](https://github.com/facebookresearch/theseus/pull/68/files/f8cf872dc3d3acffc3232eace5826c6ef486babb#diff-c5e3f8d4504d1054716f8e6877b315b545ff52580232d8433d837c74e0ef4461R35). I guess the question is if it makes sense to move the `unsqueeze` inside the `updated_from_x_y_z_quaternion` function, which I think it does.

Reply:
Let's move the `unsqueeze` for adding a batch dimension (if needed) inside `updated_from_x_y_z_quaternion` before merging.

Reply:
@fantaosha @luisenp was this addressed before merging? If not maybe add to an active PR.